src/Workspaces/Core/Portable/Indentation/AbstractIndentationService.Indenter.cs

```diff
@@ -71,48 +70,96 @@ public IndentationResult GetDesiredIndentation(FormattingOptions.IndentStyle ind
             {
                 // If the caller wants no indent, then we'll return an effective '0' indent.
                 if (indentStyle == FormattingOptions.IndentStyle.None)
+                    return default;
+
+                // If the user has explicitly set 'block' indentation, or they're in an inactive preprocessor region,
+                // then just do simple block indentation.
+                if (indentStyle == FormattingOptions.IndentStyle.Block ||
+                    _syntaxFacts.IsInInactiveRegion(Document.SyntaxTree, LineToBeIndented.Start, this.CancellationToken))
                 {
-                    return new IndentationResult(basePosition: 0, offset: 0);
+                    return GetDesiredBlockIndentation();
                 }
 
-                // find previous line that is not blank.  this will skip over things like preprocessor
-                // regions and inactive code.
-                var previousLineOpt = GetPreviousNonBlankOrPreprocessorLine();
+                Debug.Assert(indentStyle == FormattingOptions.IndentStyle.Smart);
+                return GetDesiredSmartIndentation();
+            }
 
-                // it is beginning of the file, there is no previous line exists. 
-                // in that case, indentation 0 is our base indentation.
-                if (previousLineOpt == null)
-                {
-                    return IndentFromStartOfLine(0);
-                }
+            private readonly IndentationResult GetDesiredSmartIndentation()
+            {
+                // For smart indent, we generally will be computing from either the previous token in the code, or in a
+                // few special cases, the previous trivia.
+                var token = TryGetPrecedingVisibleToken();
+
+                // Look to see if we're immediately following some visible piece of trivia.  There may
+                // be some cases where we'll base our indent off of that.  However, we only do this as 
+                // long as we're immediately after the trivia.  If there are any blank lines between us
+                // then we consider that unimportant for indentation.
+                var trivia = TryGetImmediatelyPrecedingVisibleTrivia();
+
+                if (token == null && trivia == null)
+                    return default;
+
+                return _service.GetDesiredIndentationWorker(this, token, trivia);
```

Comment:
this is the thrust of hte change.  The previous algorithm attempted to look at lines, walking backwards with some heuristics to try to find the appropriate token/trivia to work with.  The new algorithm is much simpler.  We simply grab the preceding token using bog-standard tree operations, and we also grab the preceding visible trivia (but only if we're on the line immediately after).  These two pieces of data are enough for both languages to determine how they want to indent.